File: 239_sliding_window_maximum.py
class Monoque:
    def __init__(self):
        self.arr = []
        
    def push(self, e):
        while self.arr and e > self.arr[-1]:
            self.arr.pop(-1)
        self.arr.append(e)
    
    def pop(self):
        return self.arr[0]
    
    def erase(self, e):
        try:
            self.arr.remove(e)
        except:
            pass
        
    def len(self):
        return len(self.arr)
    
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        que = Monoque()
        res = []
        for i in range(len(nums)):
            que.push(nums[i])
            # print(que.arr)
            if i-k+1 >= 0:
                res.append(que.pop())
                if que.pop()==nums[i-k+1]:
                    que.erase(nums[i-k+1])
        return res
                

File: 239_sliding_window_maximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        vector<int> res;
        multiset<int> st;
        for (int i = 0; i < nums.size(); i++){
            if (st.size() >= k) {
                // printf("%d, %d\n",i,nums[i-k]);
                st.erase(st.find(nums[i-k]));
            }
            // printf("%d st ", i);
            // for (std::multiset<int>::const_iterator j(st.begin()), end(st.end());j != end; ++j)
            //     printf("%d ", *j);
            // printf("\n");
            st.insert(nums[i]);
            if (i >= k-1)
                res.push_back(*st.rbegin());
        }
        return res;
    }
};